content/blog/2019/02/2019-02-21-credentials-masking.adoc

```diff
@@ -0,0 +1,88 @@
+This post is intended to address one of those:
+The goal and limitations of credentials masking.
+
+== The Problem
+
+One very common example of that is the role of credentials masking in Jenkins, typically involving a pipeline snippet that looks like this:
+
+[pipeline]
+----
+// Scripted //
+withCredentials([usernamePassword(credentialsId: 'topSecretCredentials', passwordVariable: 'PWD', usernameVariable: 'USR')])
+  sh './deploy.sh' // requires PWD and USR to be set
+}
+// Declarative //
+----
+
+Credentials that are in scope are made available to the pipeline without limitation.
+To prevent accidental exposure in the build log, credentials are masked from regular output, so an invocation of `env` (Linux) or `set` (Windows), or programs printing their environment or parameters would not reveal them in the build log to users who would not otherwise have access to the credentials.
+
+The misconception here is that Jenkins will prevent other, perhaps deliberate ways to reveal the password.
+Some examples:
+
+[pipeline]
+----
+// Scripted //
+withCredentials([usernamePassword(credentialsId: 'topSecretCredentials', passwordVariable: 'PWD', usernameVariable: 'USR')])
+  sh 'echo $PWD | base64' // will print e.g. dDBwczNjcjN0Cg= which is trivially converted back to the top secret password
+}
+// Declarative //
+----
+
+[pipeline]
+----
+// Scripted //
+withCredentials([usernamePassword(credentialsId: 'topSecretCredentials', passwordVariable: 'PWD', usernameVariable: 'USR')])
+  sh 'echo $PWD > myfile'
+  archiveArtifacts 'myfile' // then browse archived artifacts from the Jenkins UI
+}
+// Declarative //
+----
+
+Both of these snippets circumvent credentials masking in the build log, and show that people with control over the build script can use credentials in ways not necessarily intended or approved by admins.
+
+Obviously these are just the most straightforward examples illustrating the problem.
+Others could involve the proc file system, sending it to an HTTP server in response to a 401 authentication challenge, embedding it in the (otherwise legitimate) build result, etc.
+
+It would be great if Jenkins could allow the flexible use of credentials with no risk of exposing them through straightforward build script modifications, but realistically, it is impossible for Jenkins to police use of the credential by a build script without the support of a very specific environment setup (e.g. restrictive network configuration).
+
+It should also be noted that credentials aren't just at risk from users able to control the pipeline, typically by editing the `Jenkinsfile`.
+Actual build scripts invoked by pipelines, either shell scripts as in the example above, or more standard build tools such as Maven (controlled by `pom.xml`) are just as much of a risk if they are run inside a `withCredentials` block, or executing on the same agent as another block that passed such credentials.
```

Comment:
The “actual build scripts” may be gated with `readTrusted`.

The risk remains of a block which passes credentials followed on the same node by one which does not. Would be tricky, but possible, to extract secrets using an attack in this way.